File: advent of code/2022/deno/day7.ts
import { readInput } from "./utils.ts"

export function day7(test = false) {
  const input = readInput(7, test)
  const root = parseInput(input)

  const part1 = allDirs(root).map(getSize).filter((size) => size <= 100000).sum()

  const toFree = 30000000 - (70000000 - getSize(root))
  const part2 = allDirs(root).map(getSize).filter(size => size >= toFree).min()

  return [part1, part2]
}

type File = number
type Dir = { [filename: string]: File | Dir }
type Path = string[]

function isFile(dir: number | Dir): dir is number {
  return typeof dir === "number"
}

function allDirs(root: Dir, dirs: Dir[] = []): Dir[] {
  for (const entry of Object.values(root)) {
    if (!isFile(entry)) {
      dirs.push(entry)
      allDirs(entry, dirs)
    }
  }
  return dirs
}

function getSize(dir: Dir | number): number {
  if (isFile(dir)) return dir
  return Object.values(dir).reduce((acc: number, d) => acc + getSize(d), 0)
}

function parseInput(input: string) {
  const root = {}
  const cwd: Path = []

  const lines = input.split("\n")
  for (let i = 0; i < lines.length; i++) {
    let line = lines[i]
    if (line === "$ ls") {
      while (i < lines.length - 1 && !lines[i + 1].startsWith("$")) {
        line = lines[++i]
        const [size, filename] = line.split(" ")
        insert(root, [...cwd, filename], size === "dir" ? {} : +size)
      }
    } else if (line.startsWith("$ cd")) {
      const path = line.split(" ").slice(2)[0]
      path === ".." ? cwd.pop() : cwd.push(path)
    }
  }
  return root
}

function insert(root: Dir, path: Path, value: Dir | number) {
  let curr = root
  const dirPath = path.slice(0, path.length - 1)
  for (const p of dirPath) {
    curr[p] ??= {}
    curr = curr[p] as Dir
  }
  curr[path.at(-1)!] = value
}

File: advent of code/2022/ocaml/lib/day7.ml
open Core

type dir = { parent : dir option; mutable children : fs_entry list }
[@@deriving sexp]

and fs_entry = File of string * int | Dir of string * dir [@@deriving sexp]

let parse_input input : fs_entry =
  (* Skip first line since its always cd / *)
  let input = List.tl_exn input in
  let root : dir = { parent = None; children = [] } in
  let cwd : dir ref = ref root in
  List.iter input ~f:(fun line ->
      let split = String.split ~on:' ' line in
      match split with
      | [ "$"; "cd"; ".." ] -> cwd := !cwd.parent |> Option.value_exn
      | [ "$"; "cd"; dir ] -> (
          let child =
            List.find_exn !cwd.children ~f:(function
              | File _ -> false
              | Dir (name, _) -> String.equal name dir)
          in
          match child with
          | Dir (_name, d) -> cwd := d
          | _ -> failwith [%string "Invalid line: %{line}"])
      | [ "$"; "ls" ] -> ()
      | [ "dir"; dirname ] ->
          !cwd.children <-
            Dir (dirname, { parent = Some !cwd; children = [] })
            :: !cwd.children
      | [ size; filename ] ->
          let size = Int.of_string size in
          !cwd.children <- File (filename, size) :: !cwd.children
      | _ -> failwith [%string "Invalid line: %{line}"]);
  Dir ("/", root)

let rec size = function
  | File (_, size) -> size
  | Dir (_, dir) -> List.sum (module Int) ~f:size dir.children

let rec alldirs = function
  | File _ -> []
  | Dir (_, { children; _ }) as dir ->
      dir :: List.concat_map children ~f:alldirs

let name = function File (name, _) | Dir (name, _) -> name

let rec print_tree ?(depth = 0) tree =
  List.iter (List.range 0 depth) ~f:(fun _ -> print_string " ");
  match tree with
  | File (name, size) ->
      print_string "| ";
      print_endline [%string "%{name} %{size#Int}"]
  | Dir (name, dir) ->
      print_string "dir ";
      print_endline [%string "%{name}"];
      List.iter dir.children ~f:(print_tree ~depth:(depth + 1))

let solve ~(file : string) =
  let input = Utils.read_lines file |> parse_input in
  (* print_tree input; *)
  let part1 =
    alldirs input |> List.map ~f:size
    |> List.filter ~f:(fun size -> size < 100000)
    |> Utils.isum
  in
  let to_free = 30000000 - (70000000 - size input) in
  let part2 =
    alldirs input |> List.map ~f:size
    |> List.filter ~f:(fun size -> size >= to_free)
    |> List.min_elt ~compare:Int.compare
    |> Option.value_exn
  in
  (part1, part2)

let%test_unit _ =
  [%test_eq: int * int] (solve ~file:"day7-example.txt") (95437, 24933642)
